File: apps/cargo-tracker/app/cargotracker/booking/domain/ShipperExistenceChecker.scala
package cargotracker.booking.domain

import cargotracker.shared.domain.ShipperId

/** Shipper Context への ACL ポート（domain-model.md 準拠）。
  *
  * Booking Context は Shipper Context のエンティティに直接依存せず、本ポート経由で 存在確認のみを行う。実装は infrastructure 層で `ShipperRepository` をラップする。
  */
trait ShipperExistenceChecker:

  /** 指定 ShipperId の荷主が存在するか確認する。 */
  def exists(shipperId: ShipperId): Boolean

File: docs/article/source/csharp/apps/cargo-tracker/src/CargoTracker.Web/Booking/Infrastructure/Services/ShipperExistenceChecker.cs
using CargoTracker.Booking.Application.Internal.OutboundServices;
using CargoTracker.Booking.Application.Internal.Services;
using CargoTracker.Shared.Application.Persistence;
using CargoTracker.Shared.Domain.Model;
using Dapper;

namespace CargoTracker.Booking.Infrastructure.Services;

/// <summary>shipper テーブルだけを参照する ACL 実装。</summary>
public sealed class ShipperExistenceChecker(IDbConnectionFactory connectionFactory) : IShipperExistenceChecker
{
    public async Task<bool> ExistsAsync(ShipperId id, CancellationToken ct = default)
    {
        var surrogateId = ShipperIdCodec.ToSurrogateId(id);
        using var connection = connectionFactory.Create();
        var count = await connection.ExecuteScalarAsync<long>(new CommandDefinition(
            "SELECT COUNT(1) FROM shipper WHERE id = @Id",
            new { Id = surrogateId }, cancellationToken: ct));
        return count > 0;
    }
}
